### PR TITLE
[Keras] Surface up batch matmul as a Keras backend function

### PR DESCRIPTION
**Pull request by soumyac1999**
**Description of changes:**
Surfaces up BatchMatmul op as Keras layer. Need to discuss where to place this layer before merging in.

**Related Issues:**

Linked Issues:
- Issue #365

Issues closed by this PR:
- Closes None

**Before merging:**

- [ ] Did you update the [flexflow-third-party](https://github.com/flexflow/flexflow-third-party) repo, if modifying any of the Cmake files, the build configs, or the submodules?

### CODE REVIEW

**Comment by eddy16112** [comment]
Looks good to me, but I do not think there is a `BatchMatmul` in Keras.

**Comment by soumyac1999** [comment]
Yes, Keras has `dot` and a `batch_dot` in `keras.backend` (https://github.com/keras-team/keras/blob/727f1f3106312ca53b5c09f77087dafaaa25da0a/keras/backend.py#L2392). If we want to be faithful to the Keras API, should I add these functions instead?

For implementation, we would still need to make a layer, right? And just expose a function which creates the layer like `add` function and `Add` layer.

**Comment by eddy16112** [comment]
I think `batch_dot` is equivalent to `BatchMatmul`, right? If yes, it is better to implement `batch_dot` instead of `BatchMatmul`, just to make our API identical with the Keras API. 

Regarding the implementation, yes, you will need to make a layer. You will also need to make a `_connect_layer_2_input_1_output` instead of using the `_connect_layer_n_input_1_output`, because the latter one takes the input tensors as a list, but `dot` and `batch_dot` treat two input tensors as two arguments. 
https://github.com/flexflow/FlexFlow/blob/master/python/flexflow/keras/layers/base_layer.py

File: python/flexflow/keras/backend/__init__.py [hunk 1]
```diff
@@ -16,6 +16,7 @@
 from __future__ import absolute_import
 from __future__ import print_function
 import sys
+from .backend_functions import batch_dot, BatchMatmul
```
Comment (lockshaw) [inline python/flexflow/keras/backend/__init__.py:19]: Is the current preference in the codebase absolute or relative imports? We should max the existing python style when possible

File: python/flexflow/keras/models/base_model.py [hunk 2]
```diff
@@ -18,6 +18,7 @@
 
 from .tensor import Tensor
 from flexflow.keras.layers import Conv2D, Pooling2D, Flatten, Dense, Activation, Concatenate, Add, Subtract, Multiply, Dropout, BatchNormalization, Embedding, Reshape
+from flexflow.keras.backend import BatchMatmul
```
Comment (lockshaw) [inline python/flexflow/keras/models/base_model.py:21]: Does `BatchMatmul` exist in `keras.backend`?

Comment (soumyac1999) [reply python/flexflow/keras/models/base_model.py:21]: No. Need to rename it to `batch_dot`

**Comment by lockshaw** [comment]
@soumyac1999 ETA of finishing and merging this?

**Comment by soumyac1999** [comment]
> @soumyac1999 ETA of finishing and merging this?

I am not sure what to do with the `BatchMatmul` class. Keras only has the `batch_dot` function but we seem to need a class too (I don't know if we can get around this).

We only need to expose `batch_dot` and keep `BatchMatmul` (or some other class) for internal use. Any ideas on how to do this?

Keras `batch_dot`: https://github.com/keras-team/keras/blob/db667716ac6be59a8495b58e0d2e246e9c031599/keras/backend.py#L2462

**Comment by lockshaw** [comment]
> > @soumyac1999 ETA of finishing and merging this?
> 
> I am not sure what to do with the `BatchMatmul` class. Keras only has the `batch_dot` function but we seem to need a class too (I don't know if we can get around this).
> 
> We only need to expose `batch_dot` and keep `BatchMatmul` (or some other class) for internal use. Any ideas on how to do this?
> 
> Keras `batch_dot`: https://github.com/keras-team/keras/blob/db667716ac6be59a8495b58e0d2e246e9c031599/keras/backend.py#L2462

Don't import `BatchMatmul` into the `backend/__init__.py`, and move `BatchMatmul` into a `backend.internal` module?

File: python/flexflow/keras/backend/__init__.py [hunk 3]
```diff
@@ -16,6 +16,8 @@
 from __future__ import absolute_import
 from __future__ import print_function
 import sys
+from .backend_functions import batch_dot
+import .internal
```
Comment (lockshaw) [inline python/flexflow/keras/backend/__init__.py:20]: Why impoert `.internal`?

Comment (soumyac1999) [reply python/flexflow/keras/backend/__init__.py]: Removed

**Review by lockshaw — APPROVED** [review]

**Pull request merged**